staging/src/k8s.io/client-go/tools/leaderelection/example/main.go

```diff
@@ -0,0 +1,106 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"log"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+	"k8s.io/klog"
+)
+
+// main demonstrates a leader elected process that will step down if interrupted.
+func main() {
+	klog.InitFlags(nil)
+	flag.Parse()
+	args := flag.Args()
+	if len(args) != 3 {
+		log.Fatalf("requires three arguments: ID NAMESPACE CONFIG_MAP_NAME (%d)", len(args))
+	}
+
+	// leader election uses the Kubernetes API by writing to a ConfigMap or Endpoints
+	// object. Conflicting writes are detected and each client handles those actions
+	// independently.
+	var config *rest.Config
+	var err error
+	if kubeconfig := os.Getenv("KUBECONFIG"); len(kubeconfig) > 0 {
+		config, err = clientcmd.BuildConfigFromFlags("", kubeconfig)
+	} else {
+		config, err = rest.InClusterConfig()
+	}
+	if err != nil {
+		log.Fatalf("failed to create client: %v", err)
+	}
+
+	// we use the ConfigMap lock type since edits to ConfigMaps are less common
+	// and fewer objects in the cluster watch "all ConfigMaps" (unlike the older
+	// Endpoints lock type, where quite a few system agents like the kube-proxy
+	// and ingress controllers must watch endpoints).
+	id := args[0]
+	lock := &resourcelock.ConfigMapLock{
+		ConfigMapMeta: metav1.ObjectMeta{
+			Namespace: args[1],
+			Name:      args[2],
+		},
+		Client: kubernetes.NewForConfigOrDie(config).CoreV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity: id,
+		},
+	}
+
+	// use a Go context so we can tell the leaderelection code when we
+	// want to step down
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// listen for interrupts or the Linux SIGTERM signal and cancel
+	// our context, which the leader election code will observe and
+	// step down
+	ch := make(chan os.Signal, 1)
+	signal.Notify(ch, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		<-ch
+		log.Printf("Received termination, signaling shutdown")
+		cancel()
+	}()
+
+	// start the leader election code loop
+	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+		Lock: lock,
+		// IMPORTANT: you MUST ensure that any code you have that
+		// is protected by the lease must terminate **before**
+		// you call cancel. Otherwise, you could have a background
+		// loop still running and another process could
+		// get elected before your background loop finished, violating
+		// the stated goal of the lease.
```

Comment:
This seems like a fairly hard guarantee to make.  Do we have an example, or are you planning a follow-on PR to make this easier?

Reply:
There's a couple of options

1. most controllers are reentrant, so it doesn't actually matter (the ones I would be concerned with in controller manager are the ones that talk to external systems and allocate things, and even they are probably ok)
2. have a global "disable client" option when you get a client (basically a .Stop() on client) which is pretty easy to make work - this would be the best option for most people
3. some controllers are correctly written even today to drain, it's reasonably easy to test them

The ones who would consume this first are things like 1.  I need to look at 2.

Reply:
I implemented 2 - it should be fairly easy to use (I made the example demonstrate it)